pkg/network/message.go

```diff
@@ -65,46 +88,37 @@ const (
 
 // NewMessage returns a new message with the given payload.
 func NewMessage(cmd CommandType, p payload.Payload) *Message {
-	var (
-		size uint32
-	)
-
-	if p != nil {
-		buf := io.NewBufBinWriter()
-		p.EncodeBinary(buf.BinWriter)
-		if buf.Err != nil {
-			panic(buf.Err)
-		}
-		b := buf.Bytes()
-		size = uint32(len(b))
-	}
-
 	return &Message{
 		Command: cmd,
-		Length:  size,
 		Payload: p,
+		Flags:   None,
 	}
 }
 
 // Decode decodes a Message from the given reader.
 func (m *Message) Decode(br *io.BinReader) error {
+	m.Flags = MessageFlag(br.ReadB())
 	m.Command = CommandType(br.ReadB())
-	m.Length = br.ReadU32LE()
-	if br.Err != nil {
-		return br.Err
+	m.compressedPayload = br.ReadVarBytes()
+	if len(m.compressedPayload) > PayloadMaxSize {
+		return errors.New("invalid payload size")
 	}
-	// return if their is no payload.
-	if m.Length == 0 {
-		return nil
-	}
-	return m.decodePayload(br)
+	return m.decodePayload()
```

Comment:
You need to check `br.Err` before doing this. This is not a `DecodeBinary`, this method returns an error and it should return a proper one for any kind of message given.